Add ICAL Download Support

JishnuT09 opened the pull request:
This pull request adds a new feature to allow users to download their schedule as an iCal calendar file. The main changes include adding a "Download Calendar" button to the UI, implementing logic to extract schedule data from the table, and generating a valid iCal file for download.

JishnuT09 committed:
Add Google Calendar save button (NOT FUNCTIONAL)

Added a 'Save to Google Calendar' button to the UI.

JishnuT09 committed:
Replace Google Calendar save with iCal placeholder

Changed the 'Save to Google Calendar' button to trigger a new downloadIcal() function instead of Google API integration. Added an email input field and a placeholder alert for future iCal export functionality.

JishnuT09 committed:
Add iCal download feature for schedule

Implemented functionality to export the schedule as an iCal (.ics) file and download it from the interface. Added ScheduleItem class and supporting functions to extract schedule data and generate iCal events. Updated UI text to reflect the new download feature and removed unused email input.

JishnuT09 committed:
Refactor schedule data extraction for iCal export

geosir commented on contributors.html:
```diff
@@ -28,6 +28,8 @@ <h1>Contributors</h1>
     </li>
     <li>Anmol Singh &lt;<a href="mailto:asingh1@imsa.edu">asingh1@imsa.edu</a>&gt;
     </li>
+    <li>Jishnu Thakkar <<a href="jthakkar@imsa.edu">jthakkar@imsa.edu</a>>
```
`<` and `>` are reserved HTML characters. Please use the HTML escape characters `&lt;` to show a `<` and `&gt;` to show a `>` (see rows above).

geosir commented on index.html:
```diff
@@ -289,6 +290,12 @@ <h2>Your Schedule</h2>
                 onClick="printDiv()">
             Print
         </button>
+        <button id="save-button"
```
Could you please add `#save-button` to the style in scheduler.css used by the other buttons? (Or better yet, modify that style to apply to those buttons using a class, since it's so common?)

geosir commented on scheduler.js:
```diff
@@ -399,3 +409,166 @@ function invert(obj) {
 }
+
+// Extract schedule data from the table for iCal conversion
+getScheduleData = function () {
```
Nice work reverse-engineering the schedule table!

For maintainability, I'd like an alternative to relying on the table structure (directly reading the `<b>`, `<u>` and `<p>` tags), since this could break if someone updates the structure of the table in the future. Two suggestions off the top of my head:

1. Class data is stored in the `data-schedule` attribute of each cell, and could be extracted from there instead.

2. (Better Design) The new `scheduleData` data structure makes a lot of sense, adding a central place to store a representation of the schedule and making it easier for future contributors to work with the data. First, consider changing `scheduleData` from an array into an object (dictionary) so that classes can be looked up efficiently if they need to be modified/deleted. Then, you could add logic to the `addToTable()` and `removeFromTable()` functions that directly updates `scheduleData` from the input. Then, when `downloadIcal()` is called, `scheduleData` will already be the latest data, and there will be no need to extract the schedule again.

geosir commented on scheduler.js:
```diff
@@ -5,6 +5,17 @@ function getParameterByName(name) {
 
+class ScheduleItem {
+    constructor(time, className, room, teacher, daysOfWeek, durationMinutes) {
+        this.time = time;
```
Although classes might (currently?) be the same mod each day, it's a little confusing that it is possible to make a schedule like the one below, but the iCal invite creates events with times all equal to the first class (8am). Could this class definition be modified to represent class times as mod/day pairs? Then, the iCal code could create an event for each mod/day pair.

(This might be easiest by modifying `addToTable` as mentioned below, since it already works with mod/day pairs for each class instance.)

<img width="799" height="456" alt="image" src="https://github.com/user-attachments/assets/1bc6ea91-c9fd-42de-8425-8040af46ec09" />

geosir commented on scheduler.js:
```diff
@@ -399,3 +409,166 @@ function invert(obj) {
+
+// Download the iCal file
+function downloadIcal() {
+    /*let email = document.getElementById("email").value;
```
Please remove this comment if it's unused.

geosir commented on index.html:
```diff
@@ -8,6 +8,7 @@
     <link rel="stylesheet" href="scheduler.css">
 
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
+    <script src="https://apis.google.com/js/api.js"></script>
```
Where is this import used?

geosir reviewed (changes requested):
Hi JishnuT09, thank you for the contribution - this is a really convenient feature!

I've just added some feedback and design suggestions. Once those have been addressed, this will be good to merge.

Please let me know if you have any questions. Thanks again!

JishnuT09 committed:
Remove unused scripts and comments. Fix HTML <> symbols

JishnuT09 committed:
Use .schedule-button class for action buttons

JishnuT09 committed:
Track scheduleData and avoid duplicate colors

JishnuT09 committed:
Refactor iCal generation and event creation

JishnuT09 committed:
Merge double mods

JishnuT09 committed:
Refresh scheduleData when updating a class

JishnuT09 commented:
@geosir Thanks so much for the feedback. I went through the code and believe I implemented all of your changes. I also added a small check that doesn't allow for multiple classes to randomly be assigned the same colors. Sorry for the long delay in my response, and let me know if there's any more improvements I can make.
